pkg/epp/flowcontrol/registry/registry.go

```diff
@@ -0,0 +1,848 @@
+// # Role: The Central Orchestrator
+//
+// The `FlowRegistry` is the single source of truth for all configuration and the lifecycle manager for all shards and
+// flows. It is responsible for complex, multi-step operations such as flow registration, dynamic shard scaling, and
+// coordinating garbage collection across all shards.
+//
+// # Concurrency: The Serialized Control Plane (Actor Model)
+//
+// To ensure correctness during complex state transitions, the `FlowRegistry` employs an Actor-like pattern. All
+// administrative operations and internal state change events (e.g., GC timers, queue signals) are serialized.
+//
+// A single background goroutine (the `Run` loop) processes events from the `events` channel. This event loop acquires
+// the main `mu` lock before processing any event. External administrative methods (like `RegisterOrUpdateFlow`) also
+// acquire this lock. This strict serialization eliminates race conditions in the control plane, simplifying the complex
+// logic of the distributed state machine.
+//
+// (See package documentation in `doc.go` for the full overview of the Architecture and Concurrency Model)
+type FlowRegistry struct {
+	config *Config
+	logger logr.Logger
+
+	// clock provides the time abstraction for the registry and its components (like `gcTracker`).
+	clock clock.WithTickerAndDelayedExecution
+
+	// mu protects all administrative operations and the internal state of the registry (shard lists, `flowState`s,
+	// `nextShardID`).
+	// It ensures that these complex, multi-step operations appear atomic to the rest of the system.
+	mu sync.Mutex
+
+	// activeShards contains shards that are currently processing requests (Active).
+	// We use a slice to maintain deterministic ordering, which is crucial for configuration partitioning.
+	activeShards []*registryShard
+
+	// drainingShards contains shards that are being gracefully shut down (Draining or Drained).
+	drainingShards []*registryShard
+
+	// nextShardID is a monotonically increasing counter used to generate unique, stable IDs for shards throughout the
+	// lifetime of the process.
+	nextShardID uint64
+
+	// flowStates tracks the desired state and GC state of all flows, keyed by flow ID.
+	flowStates map[string]*flowState
+
+	// gc is the decoupled timer manager.
+	gc *gcTracker
+
+	// events is a channel for all internal state change events from shards and queues.
+	// This channel drives the serialized event loop (the control plane).
+	//
+	// CRITICAL: This is a buffered channel. Sends to this channel MUST NOT be dropped, as the GC system relies on
+	// exactly-once delivery of edge-triggered events. If this buffer fills, the data path (which sends the events) WILL
+	// block, applying necessary backpressure to the control plane.
+	events chan event
+
+	// Globally aggregated statistics.
+	totalByteSize atomic.Uint64
+	totalLen      atomic.Uint64
+	// perPriorityBandStats stores *bandStats, keyed by priority (uint).
+	perPriorityBandStats sync.Map
```

Comment:
This map is populated once at initialization and then only read from in the `Stats()` method. 

We could consider replacing with a standard map protected by the registry's mu lock (or an RWMutex for read access) which is more conventional and potentially more performant as `sync.Map` is optimized for a different access pattern (concurrent reads and writes).

This would simplify the code a bit as well.

The only writes we are concerned with are key additions and deletions. The values are already `atomic.Uint64`.

Reply:
Closing the loop on this thread: The design has since evolved to use a simple `map[uint]*bandStats`.
The map's keyset is immutable after initialization, and the values (`*bandStats`) contain atomic counters, which are updated lock-free. This approach is simpler and more idiomatic than the original `sync.Map` for our specific access pattern.